Update stub for socket module

jirassimok opened the pull request:
Summary of changes:

- Add new constants and methods from Python 3.7 and 3.8
- Add some missing Windows-specific functions
- Reorganize the many constants to mirror the documentation, including platform and version checks where possible
  - Some of the version checks may appear repetitive because I grouped the constants as [the documentation](https://docs.python.org/3.8/library/socket.html#constants) does when possible, rather than grouping by version or platform.
- Fix a few incorrect types

There are a few issues left with the socket module that I did not address:
- Many of the constants are system-specific. I did not include platform conditions for any constants without explicit availability notes in the documentation or source.
  - `AF_LINK` is listed as available on BSD and macOS. I left a comment for this, but no platform condition.
  - Some consants are documented as available on Windows only after a specific Python -version; I left these un-noted.
- There are a handful of uses of the redundant `Union[bytearray, bytes, int]` instead of `Union[bytes, int]`.
- Most uses of `_Address` are in `Union[_Address, bytes]`. I believe that `_Address` should be redefined as `Union[tuple, Text, bytes]` instead of `Union[tuple, str]`.
- _Many_ positional-only parameters are not defined as such.
- [`sock.sendmsg_afalg`](https://docs.python.org/3.8/library/socket.html#socket.socket.sendmsg_afalg) (which I added to the stub) does not have its parameter types documented
- [`sock.sendfile`](https://docs.python.org/3.8/library/socket.html#socket.socket.sendfile) (which I added to the stub) does not list a type for its first parameter (`file`)
- `socket.socket.family` and `socket.socket.type` have type `AddressFamily` and `SocketKind`, respectively, unless an unknown address family or socket type are used. They could be documented as unions with `int`, but that seems to gain little.

srittau commented on stdlib/2and3/socket.pyi:
```diff
@@ -529,8 +578,9 @@ class socket:
+    def dup(self) -> socket: ...
     def fileno(self) -> int: ...
-
+    def get_inheritable(self) -> bool: ...
```
Only available since Python 3.4 (so not in Python 2). Should have a version check.

srittau commented on stdlib/2and3/socket.pyi:
```diff
@@ -539,71 +589,101 @@ class socket:
+                 buffering: int = ...,
+                 encoding: str = ...,
+                 errors: str = ...,
+                 newline: str = ...) -> Any: ...
```
In Python 2, this only has two arguments, in Python 3 all other arguments are keyword-only. Also, all arguments except `mode` can be `None` in Python 3.

srittau commented on stdlib/2and3/socket.pyi:
```diff
@@ -539,71 +589,101 @@ class socket:
+    if sys.version_info >= (3, 5):
+        # TODO determine legal types for file parameter
+        def sendfile(self, file, offset: int = ..., count: Optional[int] = ...) -> int: ...
+    def set_inheritable(self, inheritable: bool) -> None: ...
```
Only in Python 3.

srittau commented on stdlib/2and3/socket.pyi:
```diff
@@ -626,13 +703,13 @@ def inet_aton(ip_string: str) -> bytes: ...  # ret val 4 bytes in length
-    def sethostname(name: str) -> None: ...
+def getdefaulttimeout() -> Optional[float]: ...
+def setdefaulttimeout(timeout: Optional[float]) -> None: ...
+def sethostname(name: str) -> None: ...
```
`sethostname()` is not availabe in Python 2, according to the docs.

jirassimok commented:
@srittau I've fixed all of the mistakes you noted.

I've also limited `SocketIO` to Python 3, and added `makefile`'s return types and mode flags.

(The original commits with separate fixup commits are in [this branch](https://github.com/jirassimok/typeshed/tree/pre-rebase-fixups).)

jirassimok commented:
The first time, I just forgot to run the tests locally. Then I forgot two more times.

jirassimok commented:
@srittau Any updates on this?

srittau commented on stdlib/2and3/socket.pyi:
```diff
@@ -539,71 +595,116 @@ class socket:
+def fromfd(fd: int, family: int, type: int, proto: int = ...) -> socket: ...
+
+if sys.platform == 'win32':
+    def fromshare(data: bytes) -> socket: ...
```
Python 3.3+

jirassimok commented:
I've made the two changes you noted, moved a few of the constants around,<sup>1</sup> and added a docstring noting that it's organized (and version-checked) according to the documentation.

(The old commits and pre-rebase fixup commits are in [this branch](https://github.com/jirassimok/typeshed/tree/fixup-round-2-pre-rebase).)

I could also merge #3496 (fixing `socket.socket.ioctl`) into this, if you'd like.

---
<sup>1</sup> Just making them closer to documentation order and grouping all the bluetooth constants together (and adding a comment noting the constants' order).
